{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2029a36b-3fb1-4ae2-8e1e-9e062606118c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Required packages.\n",
    "\n",
    "import gzip\n",
    "from tqdm import tqdm\n",
    "from collections import Counter\n",
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np\n",
    "import itertools\n",
    "import re\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from statistics import mean\n",
    "import random\n",
    "from collections import OrderedDict\n",
    "import pysam\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "71b493de-e8be-4d13-b1e4-3b61d2f69ceb",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# .cram - file source = HG00404.\n",
    "\n",
    "# samfile = pysam.AlignmentFile(\"../../data_pysam/HG00405.final.cram\", \"rc\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "a724823c-6cd2-4b17-ad32-ea42f2b6f90c",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# for read in samfile.fetch('chr1', 5942587, 5942604):\n",
    "    \n",
    "#     if read.mapping_quality < 5: continue\n",
    "#     if 'I' in list(read.cigarstring):\n",
    "#         print(read.cigartuples, sum(read.get_aligned_pairs(with_seq=True), ()))\n",
    "#         break\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "id": "70718742-6282-40e6-947d-9e13610fd9c5",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# # Finding out the way to get the CIGAR string pos in read, reference & type of variation.\n",
    "# # Then figuring out the way to include the way to fecth the info in our set of reads.\n",
    "\n",
    "# # pysam package or module.\n",
    "\n",
    "# set_alleles = [] \n",
    "          \n",
    "# repeat_start = 3830776\n",
    "# repeat_end = 3830795\n",
    "\n",
    "# for read in samfile.fetch('chr1',repeat_start ,repeat_end):\n",
    "    \n",
    "#     rep_len = repeat_end - repeat_start\n",
    "    \n",
    "#     if read.mapping_quality >= 5 :\n",
    "    \n",
    "    \n",
    "#         cigar_s = read.cigarstring\n",
    "#         cigar_t = sum(read.cigartuples, ())\n",
    "\n",
    "\n",
    "#         num_bases = list(cigar_t)[1::2]\n",
    "#         alignment_type = list(cigar_t)[::2]\n",
    "#         current_ref_pos = read.reference_start\n",
    "\n",
    "\n",
    "#         # saving Insertions and deletions counts which are part of repeat region in the list.  \n",
    "#         change_in_len_insert = {'1':0}\n",
    "#         change_in_len_del = []\n",
    "\n",
    "\n",
    "#         print(num_bases, alignment_type)\n",
    "#         for i in range(len(alignment_type)):\n",
    "\n",
    "\n",
    "\n",
    "#             # Here we get the actual reference pos by substracting the reference pos with hard or soft clip bases.\n",
    "#             if alignment_type[i] == 1:\n",
    "#                 # adding the insertion counts to the dictionary change_in_len\n",
    "#                 if current_ref_pos in range(repeat_start, repeat_end+1):\n",
    "#                     change_in_len_insert['1'] += num_bases[i]\n",
    "\n",
    "#             elif alignment_type[i] == 2:\n",
    "#                 # adding the deletion counts to the dict.\n",
    "#                 del_end_pos = current_ref_pos + num_bases[i]-1\n",
    "#                 l = [(change_in_len_del.append(1)) for del_pos in range(current_ref_pos, del_end_pos+1) if del_pos in range(repeat_start, repeat_end+1)] \n",
    "#                 current_ref_pos += num_bases[i]\n",
    "\n",
    "#             elif alignment_type[i] not in [4,5]:\n",
    "#                 current_ref_pos += num_bases[i]\n",
    "\n",
    "\n",
    "\n",
    "#         # change in the len of read overlapping the repeat.\n",
    "#         len_diff_read = int(change_in_len_insert['1']) - sum(change_in_len_del)\n",
    "#         Allele_From_read = rep_len + len_diff_read\n",
    "#         set_alleles.append(Allele_From_read)\n",
    "\n",
    "#         print(read.cigarstring, change_in_len_insert, change_in_len_del, rep_len, rep_len + len_diff_read, Allele_From_read, len_diff_read)\n",
    "#         print('-------------------------------------', 'Repeat_corridinates:',repeat_start, repeat_end,'|','read_corridinates:', read.reference_start, read.reference_end)\n",
    "\n",
    "# print(set(set_alleles))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "a4b050e1-8493-46af-938b-87261833257c",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true,
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "coverage at base 1186791 = 1\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = C 0\n",
      "\n",
      "coverage at base 1186792 = 1\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = A 1\n",
      "\n",
      "coverage at base 1186793 = 1\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = A 2\n",
      "\n",
      "coverage at base 1186794 = 1\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = A 3\n",
      "\n",
      "coverage at base 1186795 = 1\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = C 4\n",
      "\n",
      "coverage at base 1186796 = 1\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = G 5\n",
      "\n",
      "coverage at base 1186797 = 1\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = C 6\n",
      "\n",
      "coverage at base 1186798 = 1\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = T 7\n",
      "\n",
      "coverage at base 1186799 = 1\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = G 8\n",
      "\n",
      "coverage at base 1186800 = 1\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = T 9\n",
      "\n",
      "coverage at base 1186801 = 2\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = G 10\n",
      "\n",
      "coverage at base 1186802 = 2\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = G 11\n",
      "\tbase in read A00297:158:HT275DSXX:1:1456:19126:28541 = G 11\n",
      "\n",
      "coverage at base 1186803 = 2\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = C 12\n",
      "\n",
      "coverage at base 1186804 = 2\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = T 13\n",
      "\tbase in read A00297:158:HT275DSXX:1:1456:19126:28541 = T 13\n",
      "\n",
      "coverage at base 1186805 = 2\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = G 14\n",
      "\n",
      "coverage at base 1186806 = 2\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = T 15\n",
      "\tbase in read A00297:158:HT275DSXX:1:1456:19126:28541 = T 15\n",
      "\n",
      "coverage at base 1186807 = 2\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = T 16\n",
      "\n",
      "coverage at base 1186808 = 2\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = C 17\n",
      "\tbase in read A00297:158:HT275DSXX:1:1456:19126:28541 = C 17\n",
      "\n",
      "coverage at base 1186809 = 2\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = T 18\n",
      "\tbase in read A00297:158:HT275DSXX:1:1456:19126:28541 = T 18\n",
      "\n",
      "coverage at base 1186810 = 2\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = T 19\n",
      "\n",
      "coverage at base 1186811 = 2\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = T 20\n",
      "\tbase in read A00297:158:HT275DSXX:1:1456:19126:28541 = T 20\n",
      "\n",
      "coverage at base 1186812 = 2\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = A 21\n",
      "\n",
      "coverage at base 1186813 = 2\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = C 22\n",
      "\tbase in read A00297:158:HT275DSXX:1:1456:19126:28541 = C 22\n",
      "\n",
      "coverage at base 1186814 = 2\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = C 23\n",
      "\tbase in read A00297:158:HT275DSXX:1:1456:19126:28541 = C 23\n",
      "\n",
      "coverage at base 1186815 = 2\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = A 24\n",
      "\n",
      "coverage at base 1186816 = 2\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = T 25\n",
      "\tbase in read A00297:158:HT275DSXX:1:1456:19126:28541 = T 25\n",
      "\n",
      "coverage at base 1186817 = 3\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = C 26\n",
      "\tbase in read A00297:158:HT275DSXX:1:1456:19126:28541 = C 26\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2619:12653:10629 = C 0\n",
      "\n",
      "coverage at base 1186818 = 4\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = C 27\n",
      "\tbase in read A00297:158:HT275DSXX:1:1456:19126:28541 = C 27\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2619:12653:10629 = C 1\n",
      "\tbase in read A00404:156:HV37TDSXX:4:1176:24532:36338 = C 0\n",
      "\n",
      "coverage at base 1186819 = 4\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = A 28\n",
      "\tbase in read A00297:158:HT275DSXX:1:1456:19126:28541 = A 28\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2619:12653:10629 = A 2\n",
      "\tbase in read A00404:156:HV37TDSXX:4:1176:24532:36338 = A 1\n",
      "\n",
      "coverage at base 1186820 = 4\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = G 29\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2619:12653:10629 = G 3\n",
      "\tbase in read A00404:156:HV37TDSXX:4:1176:24532:36338 = G 2\n",
      "\n",
      "coverage at base 1186821 = 4\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = T 30\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2619:12653:10629 = T 4\n",
      "\tbase in read A00404:156:HV37TDSXX:4:1176:24532:36338 = T 3\n",
      "\n",
      "coverage at base 1186822 = 4\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = T 31\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2619:12653:10629 = T 5\n",
      "\tbase in read A00404:156:HV37TDSXX:4:1176:24532:36338 = T 4\n",
      "\n",
      "coverage at base 1186823 = 4\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = G 32\n",
      "\tbase in read A00297:158:HT275DSXX:1:1456:19126:28541 = G 32\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2619:12653:10629 = G 6\n",
      "\tbase in read A00404:156:HV37TDSXX:4:1176:24532:36338 = G 5\n",
      "\n",
      "coverage at base 1186824 = 4\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = T 33\n",
      "\tbase in read A00297:158:HT275DSXX:1:1456:19126:28541 = T 33\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2619:12653:10629 = T 7\n",
      "\tbase in read A00404:156:HV37TDSXX:4:1176:24532:36338 = T 6\n",
      "\n",
      "coverage at base 1186825 = 4\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = G 34\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2619:12653:10629 = G 8\n",
      "\tbase in read A00404:156:HV37TDSXX:4:1176:24532:36338 = G 7\n",
      "\n",
      "coverage at base 1186826 = 4\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = A 35\n",
      "\tbase in read A00297:158:HT275DSXX:1:1456:19126:28541 = A 35\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2619:12653:10629 = A 9\n",
      "\tbase in read A00404:156:HV37TDSXX:4:1176:24532:36338 = A 8\n",
      "\n",
      "coverage at base 1186827 = 4\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = A 36\n",
      "\tbase in read A00297:158:HT275DSXX:1:1456:19126:28541 = A 36\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2619:12653:10629 = A 10\n",
      "\tbase in read A00404:156:HV37TDSXX:4:1176:24532:36338 = A 9\n",
      "\n",
      "coverage at base 1186828 = 4\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = A 37\n",
      "\tbase in read A00297:158:HT275DSXX:1:1456:19126:28541 = A 37\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2619:12653:10629 = A 11\n",
      "\tbase in read A00404:156:HV37TDSXX:4:1176:24532:36338 = A 10\n",
      "\n",
      "coverage at base 1186829 = 5\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = G 38\n",
      "\tbase in read A00297:158:HT275DSXX:1:1456:19126:28541 = G 38\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2619:12653:10629 = G 12\n",
      "\tbase in read A00404:156:HV37TDSXX:4:1176:24532:36338 = G 11\n",
      "\tbase in read A00404:156:HV37TDSXX:2:2641:10465:7404 = G 0\n",
      "\n",
      "coverage at base 1186830 = 6\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = T 39\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2619:12653:10629 = T 13\n",
      "\tbase in read A00404:156:HV37TDSXX:4:1176:24532:36338 = T 12\n",
      "\tbase in read A00404:156:HV37TDSXX:2:2641:10465:7404 = T 1\n",
      "\tbase in read A00404:156:HV37TDSXX:3:1526:4218:17253 = T 0\n",
      "\n",
      "coverage at base 1186831 = 7\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = T 40\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2619:12653:10629 = T 14\n",
      "\tbase in read A00404:156:HV37TDSXX:4:1176:24532:36338 = T 13\n",
      "\tbase in read A00404:156:HV37TDSXX:2:2641:10465:7404 = T 2\n",
      "\tbase in read A00404:156:HV37TDSXX:3:1526:4218:17253 = T 1\n",
      "\tbase in read A00297:158:HT275DSXX:4:2374:11858:17989 = T 0\n",
      "\n",
      "coverage at base 1186832 = 7\n",
      "\tbase in read A00297:158:HT275DSXX:1:1456:19126:28541 = T 41\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2619:12653:10629 = T 15\n",
      "\tbase in read A00404:156:HV37TDSXX:4:1176:24532:36338 = T 14\n",
      "\tbase in read A00404:156:HV37TDSXX:2:2641:10465:7404 = T 3\n",
      "\tbase in read A00404:156:HV37TDSXX:3:1526:4218:17253 = T 2\n",
      "\tbase in read A00297:158:HT275DSXX:4:2374:11858:17989 = T 1\n",
      "\n",
      "coverage at base 1186833 = 8\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = T 42\n",
      "\tbase in read A00297:158:HT275DSXX:1:1456:19126:28541 = T 42\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2619:12653:10629 = T 16\n",
      "\tbase in read A00404:156:HV37TDSXX:4:1176:24532:36338 = T 15\n",
      "\tbase in read A00404:156:HV37TDSXX:2:2641:10465:7404 = T 4\n",
      "\tbase in read A00404:156:HV37TDSXX:3:1526:4218:17253 = T 3\n",
      "\tbase in read A00297:158:HT275DSXX:4:2374:11858:17989 = T 2\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2357:2284:12273 = T 0\n",
      "\n",
      "coverage at base 1186834 = 8\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = T 43\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2619:12653:10629 = T 17\n",
      "\tbase in read A00404:156:HV37TDSXX:4:1176:24532:36338 = T 16\n",
      "\tbase in read A00404:156:HV37TDSXX:2:2641:10465:7404 = T 5\n",
      "\tbase in read A00404:156:HV37TDSXX:3:1526:4218:17253 = T 4\n",
      "\tbase in read A00297:158:HT275DSXX:4:2374:11858:17989 = T 3\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2357:2284:12273 = T 1\n",
      "\n",
      "coverage at base 1186835 = 9\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = T 44\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2619:12653:10629 = T 18\n",
      "\tbase in read A00404:156:HV37TDSXX:4:1176:24532:36338 = T 17\n",
      "\tbase in read A00404:156:HV37TDSXX:2:2641:10465:7404 = T 6\n",
      "\tbase in read A00404:156:HV37TDSXX:3:1526:4218:17253 = T 5\n",
      "\tbase in read A00297:158:HT275DSXX:4:2374:11858:17989 = T 4\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2357:2284:12273 = T 2\n",
      "\tbase in read A00404:155:HV27LDSXX:2:2438:20772:33458 = T 0\n",
      "\n",
      "coverage at base 1186836 = 9\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = T 45\n",
      "\tbase in read A00297:158:HT275DSXX:1:1456:19126:28541 = T 45\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2619:12653:10629 = T 19\n",
      "\tbase in read A00404:156:HV37TDSXX:4:1176:24532:36338 = T 18\n",
      "\tbase in read A00404:156:HV37TDSXX:2:2641:10465:7404 = T 7\n",
      "\tbase in read A00404:156:HV37TDSXX:3:1526:4218:17253 = T 6\n",
      "\tbase in read A00297:158:HT275DSXX:4:2374:11858:17989 = T 5\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2357:2284:12273 = T 3\n",
      "\tbase in read A00404:155:HV27LDSXX:2:2438:20772:33458 = T 1\n",
      "\n",
      "coverage at base 1186837 = 9\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = T 46\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2619:12653:10629 = T 20\n",
      "\tbase in read A00404:156:HV37TDSXX:4:1176:24532:36338 = T 19\n",
      "\tbase in read A00404:156:HV37TDSXX:2:2641:10465:7404 = T 8\n",
      "\tbase in read A00404:156:HV37TDSXX:3:1526:4218:17253 = T 7\n",
      "\tbase in read A00297:158:HT275DSXX:4:2374:11858:17989 = T 6\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2357:2284:12273 = T 4\n",
      "\tbase in read A00404:155:HV27LDSXX:2:2438:20772:33458 = T 2\n",
      "\n",
      "coverage at base 1186838 = 10\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = G 47\n",
      "\tbase in read A00297:158:HT275DSXX:1:1456:19126:28541 = G 47\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2619:12653:10629 = G 21\n",
      "\tbase in read A00404:156:HV37TDSXX:4:1176:24532:36338 = G 20\n",
      "\tbase in read A00404:156:HV37TDSXX:2:2641:10465:7404 = G 9\n",
      "\tbase in read A00297:158:HT275DSXX:4:2374:11858:17989 = G 7\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2357:2284:12273 = G 5\n",
      "\tbase in read A00404:155:HV27LDSXX:2:2438:20772:33458 = G 3\n",
      "\tbase in read A00297:158:HT275DSXX:1:1563:9263:23109 = G 0\n",
      "\n",
      "coverage at base 1186839 = 11\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = T 48\n",
      "\tbase in read A00297:158:HT275DSXX:1:1456:19126:28541 = T 48\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2619:12653:10629 = T 22\n",
      "\tbase in read A00404:156:HV37TDSXX:4:1176:24532:36338 = T 21\n",
      "\tbase in read A00404:156:HV37TDSXX:2:2641:10465:7404 = T 10\n",
      "\tbase in read A00404:156:HV37TDSXX:3:1526:4218:17253 = T 9\n",
      "\tbase in read A00297:158:HT275DSXX:4:2374:11858:17989 = T 8\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2357:2284:12273 = T 6\n",
      "\tbase in read A00404:155:HV27LDSXX:2:2438:20772:33458 = T 4\n",
      "\tbase in read A00297:158:HT275DSXX:1:1563:9263:23109 = T 1\n",
      "\tbase in read A00404:156:HV37TDSXX:3:2644:11912:15139 = T 0\n",
      "\n",
      "coverage at base 1186840 = 12\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = T 49\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2619:12653:10629 = T 23\n",
      "\tbase in read A00404:156:HV37TDSXX:4:1176:24532:36338 = T 22\n",
      "\tbase in read A00404:156:HV37TDSXX:2:2641:10465:7404 = T 11\n",
      "\tbase in read A00404:156:HV37TDSXX:3:1526:4218:17253 = T 10\n",
      "\tbase in read A00297:158:HT275DSXX:4:2374:11858:17989 = T 9\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2357:2284:12273 = T 7\n",
      "\tbase in read A00404:155:HV27LDSXX:2:2438:20772:33458 = T 5\n",
      "\tbase in read A00297:158:HT275DSXX:1:1563:9263:23109 = T 2\n",
      "\tbase in read A00404:156:HV37TDSXX:3:2644:11912:15139 = T 1\n",
      "\tbase in read A00404:156:HV37TDSXX:1:2603:30888:24361 = T 0\n",
      "\n",
      "coverage at base 1186841 = 12\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2619:12653:10629 = T 24\n",
      "\tbase in read A00404:156:HV37TDSXX:4:1176:24532:36338 = T 23\n",
      "\tbase in read A00404:156:HV37TDSXX:2:2641:10465:7404 = T 12\n",
      "\tbase in read A00404:156:HV37TDSXX:3:1526:4218:17253 = T 11\n",
      "\tbase in read A00297:158:HT275DSXX:4:2374:11858:17989 = T 10\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2357:2284:12273 = T 8\n",
      "\tbase in read A00404:155:HV27LDSXX:2:2438:20772:33458 = T 6\n",
      "\tbase in read A00297:158:HT275DSXX:1:1563:9263:23109 = T 3\n",
      "\tbase in read A00404:156:HV37TDSXX:3:2644:11912:15139 = T 2\n",
      "\tbase in read A00404:156:HV37TDSXX:1:2603:30888:24361 = T 1\n",
      "\n",
      "coverage at base 1186842 = 13\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = G 51\n",
      "\tbase in read A00297:158:HT275DSXX:1:1456:19126:28541 = G 51\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2619:12653:10629 = G 25\n",
      "\tbase in read A00404:156:HV37TDSXX:4:1176:24532:36338 = G 24\n",
      "\tbase in read A00404:156:HV37TDSXX:2:2641:10465:7404 = G 13\n",
      "\tbase in read A00404:156:HV37TDSXX:3:1526:4218:17253 = G 12\n",
      "\tbase in read A00297:158:HT275DSXX:4:2374:11858:17989 = G 11\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2357:2284:12273 = G 9\n",
      "\tbase in read A00404:155:HV27LDSXX:2:2438:20772:33458 = G 7\n",
      "\tbase in read A00297:158:HT275DSXX:1:1563:9263:23109 = G 4\n",
      "\tbase in read A00404:156:HV37TDSXX:3:2644:11912:15139 = G 3\n",
      "\tbase in read A00404:156:HV37TDSXX:1:2603:30888:24361 = G 2\n",
      "\n",
      "coverage at base 1186843 = 13\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2619:12653:10629 = T 26\n",
      "\tbase in read A00404:156:HV37TDSXX:4:1176:24532:36338 = T 25\n",
      "\tbase in read A00404:156:HV37TDSXX:2:2641:10465:7404 = T 14\n",
      "\tbase in read A00404:156:HV37TDSXX:3:1526:4218:17253 = T 13\n",
      "\tbase in read A00297:158:HT275DSXX:4:2374:11858:17989 = T 12\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2357:2284:12273 = T 10\n",
      "\tbase in read A00404:155:HV27LDSXX:2:2438:20772:33458 = T 8\n",
      "\tbase in read A00297:158:HT275DSXX:1:1563:9263:23109 = T 5\n",
      "\tbase in read A00404:156:HV37TDSXX:3:2644:11912:15139 = T 4\n",
      "\tbase in read A00404:156:HV37TDSXX:1:2603:30888:24361 = T 3\n",
      "\tbase in read A00297:158:HT275DSXX:2:1121:31774:33129 = T 1\n",
      "\n",
      "coverage at base 1186844 = 13\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = T 53\n",
      "\tbase in read A00297:158:HT275DSXX:1:1456:19126:28541 = T 53\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2619:12653:10629 = T 27\n",
      "\tbase in read A00404:156:HV37TDSXX:4:1176:24532:36338 = T 26\n",
      "\tbase in read A00404:156:HV37TDSXX:2:2641:10465:7404 = T 15\n",
      "\tbase in read A00404:156:HV37TDSXX:3:1526:4218:17253 = T 14\n",
      "\tbase in read A00297:158:HT275DSXX:4:2374:11858:17989 = T 13\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2357:2284:12273 = T 11\n",
      "\tbase in read A00404:155:HV27LDSXX:2:2438:20772:33458 = T 9\n",
      "\tbase in read A00297:158:HT275DSXX:1:1563:9263:23109 = T 6\n",
      "\tbase in read A00404:156:HV37TDSXX:3:2644:11912:15139 = T 5\n",
      "\tbase in read A00404:156:HV37TDSXX:1:2603:30888:24361 = T 4\n",
      "\tbase in read A00297:158:HT275DSXX:2:1121:31774:33129 = T 2\n",
      "\n",
      "coverage at base 1186845 = 13\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = G 54\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2619:12653:10629 = T 28\n",
      "\tbase in read A00404:156:HV37TDSXX:4:1176:24532:36338 = T 27\n",
      "\tbase in read A00404:156:HV37TDSXX:2:2641:10465:7404 = T 16\n",
      "\tbase in read A00404:156:HV37TDSXX:3:1526:4218:17253 = T 15\n",
      "\tbase in read A00297:158:HT275DSXX:4:2374:11858:17989 = T 14\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2357:2284:12273 = T 12\n",
      "\tbase in read A00404:155:HV27LDSXX:2:2438:20772:33458 = T 10\n",
      "\tbase in read A00297:158:HT275DSXX:1:1563:9263:23109 = T 7\n",
      "\tbase in read A00404:156:HV37TDSXX:3:2644:11912:15139 = T 6\n",
      "\tbase in read A00404:156:HV37TDSXX:1:2603:30888:24361 = T 5\n",
      "\tbase in read A00297:158:HT275DSXX:2:1121:31774:33129 = T 3\n",
      "\n",
      "coverage at base 1186846 = 13\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = T 55\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2619:12653:10629 = T 29\n",
      "\tbase in read A00404:156:HV37TDSXX:4:1176:24532:36338 = T 28\n",
      "\tbase in read A00404:156:HV37TDSXX:2:2641:10465:7404 = T 17\n",
      "\tbase in read A00404:156:HV37TDSXX:3:1526:4218:17253 = T 16\n",
      "\tbase in read A00297:158:HT275DSXX:4:2374:11858:17989 = T 15\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2357:2284:12273 = T 13\n",
      "\tbase in read A00404:155:HV27LDSXX:2:2438:20772:33458 = T 11\n",
      "\tbase in read A00297:158:HT275DSXX:1:1563:9263:23109 = T 8\n",
      "\tbase in read A00404:156:HV37TDSXX:3:2644:11912:15139 = T 7\n",
      "\tbase in read A00404:156:HV37TDSXX:1:2603:30888:24361 = T 6\n",
      "\tbase in read A00297:158:HT275DSXX:2:1121:31774:33129 = T 4\n",
      "\n",
      "coverage at base 1186847 = 14\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = T 56\n",
      "\tbase in read A00297:158:HT275DSXX:1:1456:19126:28541 = T 56\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2619:12653:10629 = T 30\n",
      "\tbase in read A00404:156:HV37TDSXX:4:1176:24532:36338 = T 29\n",
      "\tbase in read A00404:156:HV37TDSXX:2:2641:10465:7404 = T 18\n",
      "\tbase in read A00404:156:HV37TDSXX:3:1526:4218:17253 = T 17\n",
      "\tbase in read A00297:158:HT275DSXX:4:2374:11858:17989 = T 16\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2357:2284:12273 = T 14\n",
      "\tbase in read A00404:155:HV27LDSXX:2:2438:20772:33458 = T 12\n",
      "\tbase in read A00297:158:HT275DSXX:1:1563:9263:23109 = T 9\n",
      "\tbase in read A00404:156:HV37TDSXX:3:2644:11912:15139 = T 8\n",
      "\tbase in read A00404:156:HV37TDSXX:1:2603:30888:24361 = T 7\n",
      "\tbase in read A00297:158:HT275DSXX:2:1121:31774:33129 = T 5\n",
      "\n",
      "coverage at base 1186848 = 14\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = T 57\n",
      "\tbase in read A00297:158:HT275DSXX:1:1456:19126:28541 = T 57\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2619:12653:10629 = T 31\n",
      "\tbase in read A00404:156:HV37TDSXX:4:1176:24532:36338 = T 30\n",
      "\tbase in read A00404:156:HV37TDSXX:2:2641:10465:7404 = T 19\n",
      "\tbase in read A00404:156:HV37TDSXX:3:1526:4218:17253 = T 18\n",
      "\tbase in read A00297:158:HT275DSXX:4:2374:11858:17989 = T 17\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2357:2284:12273 = T 15\n",
      "\tbase in read A00404:155:HV27LDSXX:2:2438:20772:33458 = T 13\n",
      "\tbase in read A00297:158:HT275DSXX:1:1563:9263:23109 = T 10\n",
      "\tbase in read A00404:156:HV37TDSXX:3:2644:11912:15139 = T 9\n",
      "\tbase in read A00404:156:HV37TDSXX:1:2603:30888:24361 = T 8\n",
      "\tbase in read A00297:158:HT275DSXX:2:1121:31774:33129 = T 6\n",
      "\tbase in read A00404:156:HV37TDSXX:3:2416:3703:17300 = T 1\n",
      "\n",
      "coverage at base 1186849 = 14\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = G 58\n",
      "\tbase in read A00297:158:HT275DSXX:1:1456:19126:28541 = G 58\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2619:12653:10629 = G 32\n",
      "\tbase in read A00404:156:HV37TDSXX:4:1176:24532:36338 = G 31\n",
      "\tbase in read A00404:156:HV37TDSXX:2:2641:10465:7404 = G 20\n",
      "\tbase in read A00404:156:HV37TDSXX:3:1526:4218:17253 = G 19\n",
      "\tbase in read A00297:158:HT275DSXX:4:2374:11858:17989 = G 18\n",
      "\tbase in read A00404:156:HV37TDSXX:4:2357:2284:12273 = G 16\n",
      "\tbase in read A00404:155:HV27LDSXX:2:2438:20772:33458 = G 14\n",
      "\tbase in read A00297:158:HT275DSXX:1:1563:9263:23109 = G 11\n",
      "\tbase in read A00404:156:HV37TDSXX:3:2644:11912:15139 = G 10\n",
      "\tbase in read A00404:156:HV37TDSXX:1:2603:30888:24361 = G 9\n",
      "\tbase in read A00404:156:HV37TDSXX:3:2416:3703:17300 = G 2\n",
      "\n",
      "coverage at base 1186850 = 14\n",
      "\tbase in read A00404:156:HV37TDSXX:2:1268:31557:24017 = T 60\n",
      "\tbase in read A00297:158:HT275DSXX:1:1456:19126:28541 = T 60\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_803158/3900609758.py:13: UserWarning: multiple_iterators not implemented for CRAM\n",
      "  for pileupcolumn in samfile.pileup(\"chr1\", 1186938):\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "string indices must be integers",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[99], line 19\u001b[0m\n\u001b[1;32m     14\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124mcoverage at base \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m = \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m%\u001b[39m (pileupcolumn\u001b[38;5;241m.\u001b[39mpos, pileupcolumn\u001b[38;5;241m.\u001b[39mn))\n\u001b[1;32m     15\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m pileupread \u001b[38;5;129;01min\u001b[39;00m pileupcolumn\u001b[38;5;241m.\u001b[39mpileups:\n\u001b[1;32m     17\u001b[0m             \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[38;5;124mbase in read \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m = \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m'\u001b[39m \u001b[38;5;241m%\u001b[39m\n\u001b[1;32m     18\u001b[0m                   (pileupread\u001b[38;5;241m.\u001b[39malignment\u001b[38;5;241m.\u001b[39mquery_name,\n\u001b[0;32m---> 19\u001b[0m                    \u001b[43mpileupread\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43malignment\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mquery_sequence\u001b[49m\u001b[43m[\u001b[49m\u001b[43mpileupread\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mquery_position\u001b[49m\u001b[43m]\u001b[49m), pileupread\u001b[38;5;241m.\u001b[39mquery_position)\n\u001b[1;32m     22\u001b[0m samfile\u001b[38;5;241m.\u001b[39mclose()\n",
      "\u001b[0;31mTypeError\u001b[0m: string indices must be integers"
     ]
    }
   ],
   "source": [
    "# Checking why only one of the snp allele is only overlapping with on the read, even though there are bunch of reads.\n",
    "\n",
    "# # 1st check, we use pileup to check if there is only one type of snp allele in the snp_pos among the all read depths.\n",
    "\n",
    "# # snp_pos =  5942510 [A,G] , 5942743 [G,A] , dnv_lens = 17|18, ref_len = 17\n",
    "\n",
    "# # 5942510 = 17, A, [A,G] correct------- all the reads which are supporting the snp_allele 'A' are true, there are only reads supporting the one allele 'A'.\n",
    "# # 10919012  22, G -> 22|24 - correct ----\n",
    "# #\n",
    "\n",
    "\n",
    "\n",
    "# for pileupcolumn in samfile.pileup(\"chr1\", 1186938):\n",
    "#     print(\"\\ncoverage at base %s = %s\" % (pileupcolumn.pos, pileupcolumn.n))\n",
    "#     for pileupread in pileupcolumn.pileups:\n",
    "        \n",
    "#             print('\\tbase in read %s = %s' %\n",
    "#                   (pileupread.alignment.query_name,\n",
    "#                    pileupread.alignment.query_sequence[pileupread.query_position]), pileupread.query_position)\n",
    "        \n",
    "            \n",
    "# samfile.close()\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "a4888cf4-b728-4b1e-a002-bcbc5f299045",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "3939it [01:24, 46.56it/s]\n"
     ]
    }
   ],
   "source": [
    "# Discription and details :\n",
    "\n",
    "\n",
    "with gzip.open('../../Analysis_KT/SNP_Denovo/SNPs_Intermediate_Files/SNPs_AssociatedRepeats/TRIO0001.tsv.gz', 'rt') as Trio_file:\n",
    "    \n",
    "    out = open('../../Analysis_KT/SNP_Denovo/SNPs_Intermediate_Files/SNPs_AssociatedDenovoSTR_lens/TRIO0001_SNP_AssociatedSTR_alleles.tsv', 'w')\n",
    "\n",
    "    samfile = pysam.AlignmentFile(\"../../data_pysam/HG00405.final.cram\", \"rc\")\n",
    "    \n",
    "    for debug,repeat_region in tqdm(enumerate(Trio_file)):\n",
    "        \n",
    "        repeat_region = repeat_region.strip().split('\\t')\n",
    "        #\n",
    "        start =  int(repeat_region[1])  \n",
    "        end =    int(repeat_region[2])\n",
    "        chrom = repeat_region[0]\n",
    "        \n",
    "        snp_pos = int(repeat_region[-5]) - 1\n",
    "        snp_gts = [repeat_region[-4], repeat_region[-3]]\n",
    "        snp_gts_representation = [repeat_region[-8], repeat_region[-7], repeat_region[-6]]\n",
    "        denovo_gts = repeat_region[-21].split('|')\n",
    "        repeat_len = end - start\n",
    "        \n",
    "        # here we are calling the function to store the snp allele parent of origin.\n",
    "        snp_allele_origin = snp_allele_origin_classification(snp_gts, snp_gts_representation)\n",
    "\n",
    "        # dictionary to maintian the snp_loc and repeat region enclosing the reads.\n",
    "        read_info = {}\n",
    "        \n",
    "        #\n",
    "        Read_depth_5percent = 0    \n",
    "        for read_depth,read in enumerate(samfile.fetch(chrom,start, end)):\n",
    "\n",
    "            # read quality > 5%.\n",
    "            if read.mapping_quality >= 5:\n",
    "                Read_depth_5percent += 1\n",
    "\n",
    "                Read = ''\n",
    "                if read.is_read1: Read = 'R1'\n",
    "                else: Read = 'R2'\n",
    "\n",
    "                read_query = read.query_name\n",
    "                read_start = read.reference_start\n",
    "                read_end = read.reference_end\n",
    "                base_info = sum(read.get_aligned_pairs(with_seq=True), ())\n",
    "\n",
    "                # checking if the ( read is enclosed) and ( covering the interested snp position in the flank). { R1:[s,e],[reference pos],[read base at reference pos] }\n",
    "                if not read_query in read_info: \n",
    "                    read_info[read_query] = {'R1':[], 'R2':[], 'Allele_on':[]}\n",
    "\n",
    "                    #\n",
    "                    snp_on_read_info = snp_check(read, snp_pos, Read, snp_gts, base_info)            \n",
    "                    Read_lenvar_info = STR_variation_from_read(read, start, end, repeat_len)\n",
    "                    read_info[read_query]['Allele_on'].extend(snp_on_read_info)\n",
    "                    read_info[read_query][Read].append(Read_lenvar_info)\n",
    "                    # print(snp_on_read_info)\n",
    "\n",
    "\n",
    "                else:\n",
    "                    snp_on_read_info = snp_check(read, snp_pos, Read, snp_gts, base_info)            \n",
    "                    Read_lenvar_info = STR_variation_from_read(read, start, end, repeat_len)\n",
    "                    read_info[read_query]['Allele_on'].extend(snp_on_read_info)\n",
    "                    read_info[read_query][Read].append(Read_lenvar_info)\n",
    "                    # print(snp_on_read_info)\n",
    "        \n",
    "        # Calling the function which gives Snps associated with STR alleles.\n",
    "        Snp_AssociatedSTR_GT_print = Snp_AssociatedSTR_GT(read_info, snp_gts,denovo_gts)\n",
    "        \n",
    "        # printing the required data.\n",
    "        if len(Snp_AssociatedSTR_GT_print) > 0:\n",
    "            print(repeat_len,*repeat_region[:6],repeat_region[16],f'{snp_gts[0]}|{snp_gts[1]}',f'{snp_allele_origin[0]}:{snp_allele_origin[1]};{snp_allele_origin[2]}:{snp_allele_origin[3]}', Snp_AssociatedSTR_GT(read_info, snp_gts,denovo_gts),Read_depth_5percent, read_depth+1, file=out, sep = '\\t')\n",
    "        \n",
    "        \n",
    "    samfile.close()\n",
    "    \n",
    "    out.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "75a0bf5a-8837-4b21-be6e-489fd7b2006b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Function which returns the Denovo genotype assiciating with the Snp genotype bases which further can be related to Denovo allele parent of origin.\n",
    "\n",
    "def Snp_AssociatedSTR_GT(read_info, snp_gts, denovo_gts):\n",
    "    \n",
    "    # \n",
    "    read_info_copy = read_info.copy()\n",
    "    \n",
    "    Dnv_PofO = []    \n",
    "    \n",
    "    # Before entering into the current repeat region read_info dictionary, we must ignore if ( read is not enclosed the repeat region | No snp on any read | or snp on R1 or R2 but no read overlapping the repeat region.\n",
    "    for read_id in read_info:\n",
    "        \n",
    "        if len(read_info[read_id]['Allele_on']) < 2 or len(read_info[read_id]['R1'])==0 and len(read_info[read_id]['R2'])==0 : read_info_copy.pop(read_id)\n",
    "        elif len(read_info[read_id]['R1']) == 1 and read_info[read_id]['R1'][0] == 0  and  len(read_info[read_id]['R2']) == 1 and read_info[read_id]['R2'][0] == 0 : read_info_copy.pop(read_id)\n",
    "        elif read_info[read_id]['Allele_on'].count(0) > 1: read_info_copy.pop(read_id)\n",
    "    \n",
    "    #\n",
    "    Snp_alleles_to_DnvAllele_counts = {snp_gts[0]:[], snp_gts[1]:[]} \n",
    "    \n",
    "    #\n",
    "    list_DnvAlleles = [Snp_alleles_to_DnvAllele_counts[read_info_copy[read_ID]['Allele_on'][1].upper()].extend(read_info_copy[read_ID]['R1'] + read_info_copy[read_ID]['R2']) for read_ID in read_info_copy.keys() if read_info_copy[read_ID]['Allele_on'][0] != 0]\n",
    "    list_DnvAlleles1 = [Snp_alleles_to_DnvAllele_counts[read_info_copy[read_ID]['Allele_on'][2].upper()].extend(read_info_copy[read_ID]['R1'] + read_info_copy[read_ID]['R2']) for read_ID in read_info_copy.keys() if read_info_copy[read_ID]['Allele_on'][0] == 0]\n",
    "    \n",
    " \n",
    "    # In here we are considering the major count allele associating with Snp allele.\n",
    "    for i,snp_allele in enumerate(Snp_alleles_to_DnvAllele_counts.keys()):\n",
    "        \n",
    "        #\n",
    "        snp_allele_itr = Counter(Snp_alleles_to_DnvAllele_counts[snp_allele])\n",
    "        if len(snp_allele_itr) == 0 : continue\n",
    "        \n",
    "        #\n",
    "        if int(denovo_gts[0]) in snp_allele_itr or int(denovo_gts[1]) in snp_allele_itr:\n",
    "            \n",
    "            # Deleting the 0's from the list of alleles, where 0 means that perticular read is not enclosing the repeat region.\n",
    "            if 0 in snp_allele_itr: snp_allele_itr.pop(0) \n",
    "            \n",
    "            # checking if there is only 0 as a key in dictionary which gets deleted in previous step.\n",
    "            if len(list(snp_allele_itr.values())) != 0 :\n",
    "            \n",
    "                values = list(snp_allele_itr.values())\n",
    "                keys = list(snp_allele_itr.keys())\n",
    "\n",
    "                #\n",
    "                index_max_value = values.index(max(values))            \n",
    "                Dnv_AlleleAssociated_snp = keys[index_max_value]\n",
    "\n",
    "                #\n",
    "                if Dnv_AlleleAssociated_snp != 0:\n",
    "                    Dnv_allele_depth = max(values)\n",
    "                    Dnv_PofO.append([Dnv_AlleleAssociated_snp,snp_allele,Dnv_allele_depth])# Here we want to print not just the max read supporting allele, we print list of allele which are getting printed.\n",
    "        \n",
    "        \n",
    "    \n",
    "    return Dnv_PofO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "d1df4005-3263-4b8e-863e-96d403ec021f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Creating function to check the snp_pos overlapping with which read and matching with which snp_allele\n",
    "\n",
    "def snp_check(read, snp_pos, Read, snp_gts, base_info):\n",
    "    \n",
    "    # base info = [read_pos, refn_pos, read_base]\n",
    "    reference_pos = list(base_info)[1::3]\n",
    "    read_bases = list(base_info)[2::3]\n",
    "    \n",
    "    # check if read is overalpped with snp_pos and base.\n",
    "    if snp_pos in reference_pos: # with the only confirmation that read.get_aligned_pairs() gives information like 'read_pos, reference_pos, base' for matched, deletion and substitution bases for other bases it reports None.\n",
    "        read_index = list(base_info).index(snp_pos)\n",
    "        # Here we are ignoring the deletion in snp_pos, coz as we are only considering substitutions from the snp_vcf file.\n",
    "        if base_info[read_index-1] != None: \n",
    "            read_base = list(str(read.query_sequence))[base_info[read_index-1]]\n",
    "\n",
    "            # checking if snp_pos base matchs with read base in same pos.\n",
    "            if read_base.upper() in snp_gts:\n",
    "                 return [Read, read_base]\n",
    "        \n",
    "    else: return []\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "e8260692-cedf-4e2d-8ff8-595122104da4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Considering only Repeat overlapped CIGAR variants.\n",
    "\n",
    "def STR_variation_from_read(read, start, end, repeat_len):\n",
    "\n",
    "    \n",
    "    # 1st we check if the read is enclosing the repeat or not, if not then return the empty list.\n",
    "    if read.reference_start <= int(start)  and  read.reference_end >= int(end):\n",
    "        \n",
    "        # Finding out the way to get the CIGAR string pos in read, reference & type of variation.\n",
    "        # Then figuring out the way to include the way to fecth the info in our set of reads.\n",
    "\n",
    "        repeat_start = int(start)\n",
    "        repeat_end = int(end)\n",
    "\n",
    "\n",
    "        cigar_s = read.cigarstring\n",
    "        cigar_t = sum(read.cigartuples, ())\n",
    "\n",
    "        num_bases = list(cigar_t)[1::2]\n",
    "        alignment_type = list(cigar_t)[::2]\n",
    "\n",
    "        # We update while itterating through the read positions to cover the entire read positions with any len and capture the variations which encounter within the repeat region.\n",
    "        current_ref_pos = read.reference_start\n",
    "\n",
    "        # saving Insertions and deletions counts which are part of repeat region in the list.\n",
    "        change_in_len_insert = {'1':0}\n",
    "        change_in_len_del = []\n",
    "\n",
    "        for indx in range(len(alignment_type)): # this loop runs for every read, itterating through M's I's D's.\n",
    "\n",
    "            # Here we get the actual reference pos by substracting the reference pos with hard or soft clip bases.\n",
    "            if alignment_type[indx] == 1:\n",
    "                # adding the insertion counts to the dictionary change_in_len\n",
    "                if current_ref_pos in range(repeat_start-1, repeat_end+1) :\n",
    "                    change_in_len_insert['1'] += num_bases[indx]\n",
    "\n",
    "\n",
    "            elif alignment_type[indx] == 2:\n",
    "                # adding the deletion counts to the dict.\n",
    "                del_end_pos = (current_ref_pos + num_bases[indx])\n",
    "                l = [(change_in_len_del.append(1)) for del_pos in range(current_ref_pos+1, del_end_pos+1) if del_pos in range(repeat_start, repeat_end+1)] \n",
    "                current_ref_pos += num_bases[indx]\n",
    "\n",
    "\n",
    "            elif alignment_type[indx] not in [4,5]:\n",
    "                current_ref_pos += num_bases[indx]\n",
    "\n",
    "\n",
    "        # change in the len of read overlapping the repeat.\n",
    "        Insertion_substract_Deletion = int(change_in_len_insert['1']) - sum(change_in_len_del)\n",
    "        len_diff_in_read = repeat_len + Insertion_substract_Deletion\n",
    "\n",
    "\n",
    "        return len_diff_in_read\n",
    "\n",
    "\n",
    "    else: return 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "ee511ec6-40d2-4b9e-9e6e-8d50c4da9584",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Here we relate snp alleles to the maternal and paternal side.\n",
    "\n",
    "def snp_allele_origin_classification(snp_gts, snp_gts_representation):\n",
    "    \n",
    "    Snp_origin = []\n",
    "    \n",
    "    # We have to 0 or 1 being unique in list of GT combinations.\n",
    "    offspring = snp_gts_representation[0].split('|')\n",
    "    parents = snp_gts_representation[1].split('|') + snp_gts_representation[2].split('|')\n",
    "    GT_bases = {'0':snp_gts[0], '1':snp_gts[1]}\n",
    "    \n",
    "    # Based on the '0' and '1' counts, either being count 1 where that allele indicates it is coming from criteria satisfied condition.\n",
    "    p1 =  parents.count(offspring[0])\n",
    "    p2 = parents.count(offspring[1])\n",
    "    \n",
    "    #\n",
    "    paternal = snp_gts_representation[1].split('|')\n",
    "    maternal = snp_gts_representation[2].split('|')\n",
    "\n",
    "    # To findout the 0 or 1 coming from which parent.\n",
    "    if p1 == 1 and paternal.count(offspring[0]) == 1:\n",
    "        Snp_origin.extend([GT_bases[offspring[0]], 'P', GT_bases[offspring[1]], 'M'])\n",
    "    \n",
    "    elif p1 == 1 and maternal.count(offspring[0]) == 1:\n",
    "        Snp_origin.extend([GT_bases[offspring[0]], 'M', GT_bases[offspring[1]], 'P'])\n",
    "\n",
    "    elif paternal.count(offspring[1]) == 1:\n",
    "        Snp_origin.extend([GT_bases[offspring[1]], 'P', GT_bases[offspring[0]], 'M'])\n",
    "\n",
    "    else:  Snp_origin.extend([GT_bases[offspring[1]], 'M', GT_bases[offspring[0]], 'P'])\n",
    "\n",
    "\n",
    "    return Snp_origin   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ff14789-a610-4127-8222-ebd09040c838",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "369791f7-33e4-4c5d-83d8-3d68220bdac7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1a291be-e1f1-4646-8383-b995646efc87",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
